File: cpp/cpp-roman-numerals-bidirectional-converter/code.cpp
/*
Copyright notice:
© 2025 Adrian-Gabriel Doaga. All Rights Reserved.
Email: adriangabrieldoaga@gmail.com
Website: https://github.com/adriangabrieldoaga/pieces-of-code
*/

#include <iostream>
#include <map>
#include <vector>

using std::cout;
using std::cin;
using std::string;
using std::vector;
using std::pair;
using std::map;

string int_To_Roman(int int_var)
{
    vector<pair<int, string>> Roman_chart =
    {
        {1000, "M"}, {900, "CM"}, {500, "D"}, {400, "CD"}, {100, "C"}, {90, "XC"}, {50, "L"}, {40, "XL"},
        {10, "X"}, {9, "IX"}, {5, "V"}, {4, "IV"}, {1, "I"}
    };

    string result = "";
    for (const auto &pair : Roman_chart)
    {
        while (int_var >= pair.first)
        {
            result += pair.second;
            int_var -= pair.first;
        }
    }
    return result;
}

int Roman_To_Int(string Roman_numeral)
{
    map<char, int> Roman_chart = {{'I', 1}, {'V', 5}, {'X', 10}, {'L', 50}, {'C', 100}, {'D', 500}, {'M', 1000}};
    int result = 0;
    int previous = 0;

    for (int i = Roman_numeral.length() - 1; i >= 0; i--)
    {
        int current = Roman_chart[Roman_numeral[i]];
        if (current < previous)
        {
            result -= current;
        }
        else
        {
            result += current;
        }
        previous = current;
    }
    return result;
}

int main()
{
    int choice;

    do
    {
        cout << "0. Close program\n";
        cout << "1. Integer to Roman\n";
        cout << "2. Roman to integer\n";
        cout << "3. Print all roman numerals from 1 to 3999\n";
        cout << "Choice: ";
        cin >> choice;
        if (choice == 0)
        {
            cout << "You closed the program.\n";
            break;
        }
        else
        if (choice == 1)
        {
            int int_var;
            cout << "Enter an integer (1 - 3999): ";
            cin >> int_var;
            if (int_var <= 0 || int_var > 3999)
            {
                cout << "Number out of range.\n";
            }
            else
            {
                cout << "Roman numeral: " << int_To_Roman(int_var) << "\n";
            }
            break;
        }
        else
        if (choice == 2)
        {
            string Roman_numeral;
            cout << "Enter a Roman numeral: ";
            cin >> Roman_numeral;
            cout << "Integer value: " << Roman_To_Int(Roman_numeral) << "\n";
            break;
        }
        else
        if (choice == 3)
        {
            for (int int_var = 1; int_var <= 3999; int_var++)
            {
                cout << int_var << " = " << int_To_Roman(int_var) << "\n";
            }
            break;
        }
        else
        {
            cout << "Invalid choice. Try again.\n\n";
            continue;
        }
    } while (true);

    return 0;
}

File: matlab/matlab_ballistic_trajectory/code.m
%{
Copyright notice:
© 2025 Adrian-Gabriel Doaga. All Rights Reserved.
Email: adriangabrieldoaga@gmail.com
Website: https://github.com/adriangabrieldoaga/pieces-of-code
%}

% The program plots the trajectory of a projectile launched in a homogeneous
% gravitational field with fluid friction.

clc % Clears Command Window (optional)
clear % Deletes memory of previous activity (optional)
close all % Closes all figures (optional)

% Physical quantities:
m = 1; % mass of projectile (kg)
v_0 = 320; % initial speed of the projectile (m / s)
alpha_0 = 31; % launch angle (between v0 and the horizontal axis) (°)
g = 9.81; % gravitational acceleration (m / s^2)
b = 0.75 * m * g / v_0; % fluid friction coefficient (N * s / m)

% Discretization of the time of movement
t_i = 0; % time of launch
% Overestimation of movement time, based on case b = 0:
t_f = 2 * v_0 * sind(alpha_0) / g; % time of contact with the ground
N = 2500; % discretization moments (number of steps)
t = linspace(t_i, t_f, N); % t is the array of moments of time from t_i to t_f, with constamt step

% motion rules in function of time (x and y are arrays)
x = v_0 * cosd(alpha_0) * m / b * (1 - exp(-b / m * t)); % distance (m)
y = v_0 * sind(alpha_0) * m / b * (1 - exp(-b / m * t)) - g * m / b * t + g * (m / b)^2 * (1 - exp(-b / m * t)); % altitude (m)

% We eliminate the elements where y < 0 from x, y and t
indices = (y >= 0); % The elements of indices arrat are 1 for y >= 0 and 0 otherwise
x = x(indices == 1);
y = y(indices == 1);
t = t(indices == 1);

% Plot
plot(x, y, 'b') % y in function of x; continuous blue line
daspect([1 1 1]) % units of measure on the x, y and t axes are in ratio 1:1:1
xlabel('x (m)');
ylabel('y (m)');
grid
title('Ballistic trajectory')
ylim([0, 1.1 * max(y)]) % see a bit above the graph

display_var = ['The projectile traveled a distance of ', num2str(x(length(t))), ' meters.'];
disp(display_var)

display_var = ['The projectile reached a maximum height of ', num2str(max(y)), ' meters.'];
disp(display_var)

display_var = ['The projectile was in motion for ', num2str(t(length(t))), ' seconds.', newline]; % t(length(t)) - last element in array t
disp(display_var)

%{
Output:
The projectile traveled a distance of 5964.6862 meters.
The projectile reached a maximum height of 1107.0322 meters.
The projectile was in motion for 30.1453 seconds.
%}